Compilation fail with serde v1.0.122

How to reproduce ?
* Check out this repo with latest commit: b52fc0a8270d671e33dafe8dec355624727e8534
* Run: `cargo update -p serde:1.0.122`
* Run `cargo check` in this repo

Error message:
```rust
error[E0283]: type annotations needed
   --> src/cargo/util/config/de.rs:530:63
    |
530 |                 seed.deserialize(Tuple2Deserializer(1i32, env.as_ref()))
    |                                                           ----^^^^^^--
    |                                                           |   |
    |                                                           |   cannot infer type for type parameter `T` declared on the trait `AsRef`
    |                                                           this method call resolves to `&T`
    |
    = note: cannot satisfy `std::string::String: AsRef<_>`
help: use the fully qualified path for the potential candidates
    |
530 |                 seed.deserialize(Tuple2Deserializer(1i32, <std::string::String as AsRef<OsStr>>::as_ref(env)))
    |                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
530 |                 seed.deserialize(Tuple2Deserializer(1i32, <std::string::String as AsRef<std::path::Path>>::as_ref(env)))
    |                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
530 |                 seed.deserialize(Tuple2Deserializer(1i32, <std::string::String as AsRef<str>>::as_ref(env)))
    |                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
530 |                 seed.deserialize(Tuple2Deserializer(1i32, <std::string::String as AsRef<[u8]>>::as_ref(env)))
    |                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

cc https://github.com/rust-lang/rust/pull/81327#issuecomment-766762024

Same here, `cargo-outdated` installation fails
```
2021-01-25T12:29:09.3660856Z error[E0283]: type annotations needed
2021-01-25T12:29:09.3662485Z    --> /home/vsts/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.50.0/src/cargo/util/config/de.rs:530:63
2021-01-25T12:29:09.3663348Z     |
2021-01-25T12:29:09.3663785Z 530 |                 seed.deserialize(Tuple2Deserializer(1i32, env.as_ref()))
2021-01-25T12:29:09.3664576Z     |                                                           ----^^^^^^--
2021-01-25T12:29:09.3665129Z     |                                                           |   |
2021-01-25T12:29:09.3665719Z     |                                                           |   cannot infer type for type parameter `T` declared on the trait `AsRef`
2021-01-25T12:29:09.3666558Z     |                                                           this method call resolves to `&T`
2021-01-25T12:29:09.3667021Z     |
2021-01-25T12:29:09.3667552Z     = note: cannot satisfy `std::string::String: AsRef<_>`
```

--- a/src/cargo/util/config/de.rs
+++ b/src/cargo/util/config/de.rs
@@ -527,7 +527,7 @@ impl<'de, 'config> de::MapAccess<'de> for ValueDeserializer<'config> {
                 seed.deserialize(Tuple2Deserializer(0i32, path.to_string_lossy()))
             }
             Definition::Environment(env) => {
-                seed.deserialize(Tuple2Deserializer(1i32, env.as_ref()))
+                seed.deserialize(Tuple2Deserializer(1i32, env.as_str()))
             }
             Definition::Cli => seed.deserialize(Tuple2Deserializer(2i32, "")),
         }

